icefoganalytics/travel-authorization
# Conform Travel Request List Page (`/my-travel-requests`) to Wireframe

Fixes https://github.com/icefoganalytics/travel-authorization/issues/16

References
- [Modeling, My Travel Requests, 2023-10-13](https://docs.google.com/document/d/1jpW3VokZa3v5Mv-Gf4Kfjve2WwTpK7X4cpep5-EROCE)

# Context

Update "My Travel Requests" page so that it more closely matches the new wireframe.
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/b323e0bc-80d9-4d37-856f-2fb21723dd76)

# Requirements

1. Add additional states to list of Form statuses.
2. State management for Form object, might want to use https://stately.ai/docs/quick-start or https://github.com/jakesgordon/javascript-state-machine
3. Pretty heavy Form serializer logic for table view as seen from analyis in [Modeling, My Travel Requests, 2023-10-13](https://docs.google.com/document/d/1jpW3VokZa3v5Mv-Gf4Kfjve2WwTpK7X4cpep5-EROCE/edit#heading=h.9699cg9h1tnl)
4. Components for each of the Travel Action call-to-action button/links (5 in total)
5. Dummy pages or effects for the various actions.
6. Some system to trigger a Form object to move through the various states, might be best to keep it as a "magic" development button or set of URL? Maybe have URLs to generate forms in the appropriate state?

## Out of Scope

1. Admin/Approver interface - this should be a completely different page and component tree
2. Implementation of call-to-action button/link pages and actions.

# Pre-Work

- [x] Automate database migrations
- [x] Fix up/standardize the various tables that are going to be used
- [ ] Abstract Form state management?

# Implementation Choices

1. Rename Design folder to _Design so it's at the top of the folder list. Purely a developer cosmetic.
2. Add seeds directory to knexfile config.
3. Drop front-end library `vuetify-datetime-picker` from back-end packages.json
4. Upgrade the `currentUser` returned by the `BaseController` such that it returns a Sequelize User object with the field names normalized to camelCase.
5. Add a QA module to the back-end for triggering various "QA" scenarios. This is an extension of the "seeds" concept, but the specific seeds are triggered by the end user.
6. Separate out the `count` from `findAll` in the `TravelAuthorizationsController` as complex queries will return a higher count than the actual number of records do to query joins.
7. Add a bunch more states to the TravelAuthorization model to match the requested UI for the My Travel Requests table.
8. Support nulls in some model definitions for foreign keys that are nullable.
9. Add state management restriction to only listed states for some models.
10. Add state management restriction to TravelDeskTravelRequest model.
11. Add non-normalized User model to Sequelize models.
12. Update `BaseSerializer` to use the "Template Pattern" to make it easier to write complex serializer logic via instance methods.
13. Handle most data presentation logic for the "My Travel Requests" table in the `TravelAuthorizationsSerializer`, while still using indirection for the "status", "phase", and "action" columns. The back-end returns a computer style snake_cased name, which the front-end is free to consume or changes as necessary. This avoids a dangerous coupling between the front- and back-end logic.
14. Update the `BaseService` to also use the "Template Pattern" to simplify the calling interface for services.
15. Add `api/src/services/qa/scenarios/my-travel-requests-service.ts` to handle setting up the appropriate states for the travel authorizations to match the wireframed UI.
16. Add `vue-i18n` library to handle translation patterning in the front-end. This makes it easy to write code that takes a back-end style `status` like `awaiting_director_approval` and convert it to the human readable `Awaiting Director Approval` in a standard way.
17. Add "refresh" button to My Travel Requests table so I don't have to wait for the "/unit" query to return from the back-end every time I want to update the page content. This can be removed once the `/unit` query is using appropriate caching.
18. Indirect "Travel Action" between the back- and front-end implementation by returning an action list from the back-end in the form `["view_itinerary"]` and abstracting each action into its own component.
19. Standardize front-end `travelForm` module naming to `travel-form`.
20. Fix incorrect kebab casing of slot names, bug introduced by me in https://github.com/icefoganalytics/travel-authorization/commit/0e2ae1199d0d69d9d3904f6c9ada1705696c6dae
21. Add front-end QA Scenarios page to the top nav bar menu. This takes you to a page that lets you trigger various QA scenarios. This isn't really a great solution long term, but it's necessary to do testing, without a test suite, against time dependent display logic. An alternative to this would be to add a bunch of tests, and to have documentation about how to get a "travel authorization" into a given state.

# Screenshots

Old - Travel Authorization list view (My Travel - /my-travel-requests)
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/a4916fb8-ab63-4a98-86fb-a758379ac03d)

New - Travel Authorization list view (My Travel - /my-travel-requests)
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/a6657620-e34f-4df9-afc2-17be5e57dfab)

QA Scenario page
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/735a5136-330f-4f74-b249-760abd4d4334)

# Testing Instructions

1. Boot the database and api services via `dev up`.
2. Boot the web service via `dev up_web`
3. Log in to the app at http://localhost:8080/
4. Check that you can navigate to the "My Travel Requests" page via the top nav drop-down.
5. Check that you can still create/read/update travel authorizations
6. In a new tab go to the magic http://localhost:8080/qa/scenarios endpoint, or select it via the top nav drop-down menu.
23. Trigger the "MY-Travel-Requests" QA scenario.
24. Go back to the http://localhost:8080/my-travel-requests page.
25. Check that the travel authorizations you see match the wireframe.
